### PR TITLE
Descriptor parser support

### PR DESCRIPTION
**Pull request by tlakollo**
Add DescriptorMarker support from linker
Add node to represent loading an assembly embedded file and checking for its dependencies
Verify for embedded files every time a module constructor is called
Support for events and properties in the XML reading
Add nodes to the dependency graph instead of marking logic
Add tests for embedded descriptor file in the NativeAOT smoke tests infra

### CODE REVIEW

**Review by MichalStrehovsky — COMMENTED** [review]
Looks great! Just a couple suggestions.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ModuleUseBasedDependencyAlgorithm.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+using DependencyList = ILCompiler.DependencyAnalysisFramework.DependencyNodeCore<ILCompiler.DependencyAnalysis.NodeFactory>.DependencyList;
+using CombinedDependencyList = System.Collections.Generic.List<ILCompiler.DependencyAnalysisFramework.DependencyNodeCore<ILCompiler.DependencyAnalysis.NodeFactory>.CombinedDependencyListEntry>;
+using DependencyListEntry = ILCompiler.DependencyAnalysisFramework.DependencyNodeCore<ILCompiler.DependencyAnalysis.NodeFactory>.DependencyListEntry;
+using System.Reflection;
+using System.Reflection.PortableExecutable;
+using System.IO;
+using System.Reflection.Metadata;
+using Internal.TypeSystem.Ecma;
+
+namespace ILCompiler.Compiler.Compiler.DependencyAnalysis
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ModuleUseBasedDependencyAlgorithm.cs:21]:
```suggestion
namespace ILCompiler.Compiler.DependencyAnalysis
```

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EmbeddedTrimmingDescriptorNode.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+using Internal.TypeSystem;
+using Internal.Runtime;
+using Internal.IL;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EmbeddedTrimmingDescriptorNode.cs:10]: Do we need all these usings?

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ModuleUseBasedDependencyAlgorithm.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using ILCompiler.DependencyAnalysis;
+using Internal.TypeSystem;
+
+using DependencyList = ILCompiler.DependencyAnalysisFramework.DependencyNodeCore<ILCompiler.DependencyAnalysis.NodeFactory>.DependencyList;
+using CombinedDependencyList = System.Collections.Generic.List<ILCompiler.DependencyAnalysisFramework.DependencyNodeCore<ILCompiler.DependencyAnalysis.NodeFactory>.CombinedDependencyListEntry>;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ModuleUseBasedDependencyAlgorithm.cs:13]: Could you organize these usings and remove unused?

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/NodeFactory.cs [hunk 4]
```diff
@@ -4,6 +4,7 @@
 using System;
 using System.Collections.Concurrent;
 using System.Collections.Generic;
+using System.Collections.ObjectModel;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/NodeFactory.cs:7]: Do we need this?

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/NonGCStaticsNode.cs [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 using System;
 using System.Collections.Generic;
-
+using ILCompiler.Compiler.Compiler.DependencyAnalysis;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/NonGCStaticsNode.cs:6]: These will go away once you change the namespace - please make sure to leave an empty line - we separate system namespaces although it's not enforced by the CI (there is a rule for it but we don't run formatter like linker repo).

**Review by MichalStrehovsky — APPROVED** [review]
Looks great otherwise, thanks!

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs [hunk 6]
```diff
@@ -191,6 +193,8 @@ protected virtual void ProcessTypes(ModuleDesc assembly, XPathNavigator nav, boo
                 // TODO: Process exported types
 
                 // TODO: Semantics differ and xml format is cecil specific, therefore they are discrepancies on things like nested types
+                // for now just hack replacing / for + to support basic resolving of nested types
+                fullname = fullname.Replace("/", "+");
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs:197]: Could you file an issue on this? Looks like something we would want to fix for 7.0.

Comment (tlakollo) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ProcessLinkerXmlBase.cs]: Filled https://github.com/dotnet/runtime/issues/73083

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/CodeBasedDependencyAlgorithm.cs [hunk 7]
```diff
@@ -6,13 +6,16 @@
 using Internal.IL;
 using Internal.Text;
 using Internal.TypeSystem;
+
+using ILCompiler.Compiler.DependencyAnalysis;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/CodeBasedDependencyAlgorithm.cs:10]: Sorry, looks like the namespace I suggested is still not the namespace we use elsewhere. I think this should just be Ilcompiler.DependencyAnalysis

**Pull request merged**